src/librustc/traits/util.rs

```diff
@@ -351,11 +379,15 @@ impl<'cx, 'gcx, 'tcx> TraitRefExpander<'cx, 'gcx, 'tcx> {
             })
             .collect();
 
-        debug!("expand_trait_refs: trait_ref={:?} items={:?}",
-                trait_ref, items);
+        debug!("expand_trait_refs: items={:?}", items);
 
         // Only keep those items that we haven't already seen.
-        items.retain(|i| self.visited.insert(&i.trait_ref().to_predicate()));
+        // Note: we also retain auto traits here for the purpose of linting duplicate auto traits
```

Comment:
Oh, I guess this can't be for backwards compatibility, since it only applies to trait alias expansion (which didn't exist before), right?

In that case, I think we should tweak this (see the "main review comment" for more details).

Reply:
Yeah, sounds reasonable. I was originally doing it for backwards compatibility, but under an assumption of unnecessary generality. I would have had to do it this way if trait aliases were always (ab)used in this way, but fortunately they're still unstable. :-)